nixosTest: adds support for lib.extend

baloo opened the pull request:
###### Description of changes

When lib overrides were used, before this PR, they would not be made available in the configuration evaluation of nixosTest's nodes.

Sample code:
``` nix
let
  pkgs = import <nixpkgs> {
    overlays = [
      (new: old: {
        lib = old.lib.extend (self: super: {
          sorry_dave = builtins.trace "There are no pod bay doors" "sorry dave";
        });
      })
    ];
  };
in
pkgs.testers.nixosTest {
  name = "demo lib overlay";

  nodes = {
    machine = { lib, ... }: {
      environment.etc."got-lib-overlay".text = lib.sorry_dave;
    };
  };

  testScript = { nodes }:
    ''
      start_all()
      machine.succeed('grep dave /etc/got-lib-overlay')
    '';
}
```

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

baloo committed:
nixosTest: adds support for lib.extend

When lib overrides were used, before this commit, they would not be made
available in the configuration evaluation of nixosTest's nodes.

Sample code:
``` nix
let
  pkgs = import ./. {
    overlays = [
      (new: old: {
        lib = old.lib.extend (self: super: {
          sorry_dave = builtins.trace "There are no pod bay doors" "sorry dave";
        });
      })
    ];
  };
in
pkgs.testers.nixosTest {
  name = "demo lib overlay";

  nodes = {
    machine = { lib, ... }: {
      environment.etc."got-lib-overlay".text = lib.sorry_dave;
    };
  };

  testScript = { nodes }:
    ''
      start_all()
      machine.succeed('grep dave /etc/got-lib-overlay')
    '';
}
```

baloo committed:
nixosTest: provide a test for lib.extend in nixosTests & runNixOSTest

baloo commented:
the second commit is here to provide a non-regression test, but I'm not too happy with it. If someone have a idea to improve, send away.

chreekat commented:
The code is very different from the description of the PR! You're not just bringing in lib.extend, you're bringing in all of lib. Right? Or am I reading it wrong.

I don't know whether or not inheriting lib has consequences, but I think the commit should describe what it's doing a lot more explicitly.

Plus, I don't understand the rest of the comment either. Is the example code something you couldn't do before, that works now? But the line above it is describing the way things were, rather than the way they are now? It's a bit confusing!

baloo commented:
`lib.extend` is something you can do in your overlay. And if you evaluate a system with it, it would work. But this would not carry over to the nixos evaluation used in the `nixosTests` because the non-overlaid version of lib is used.

baloo commented:
@chreekat I'm still confused by your 'code looks different than the description". could you review the first commit first, then the second one.
The second one is just here to bring a test (and it adds noise to the overall diff). But the fix here is really one line.

roberth commented on nixos/tests/nixos-test-driver/lib-extend.nix:
```diff
@@ -0,0 +1,27 @@
+        });
+      })
+    ];
+  };
```
This relies on `builtins.currentSystem` instead of `stdenv.*Platform`, so it breaks `--argstr system foo-linux -A tests.nixos-test-driver.lib-extend`.
You could perhaps use the `pkgs.extend` function to add the overlay. That seems like a simpler solution than getting the parameters and mangling them back into a `localSystem`/`crossSystem` pair. (Or we could make Nixpkgs take *Platform parameters like NixOS does nowadays...)

baloo replied:
Oh yeah you're right! I completely forgot about `pkgs.extend`! I was focused on my use-case that made use of overlays, but pkgs.extend just works the same here.

roberth commented on nixos/tests/nixos-test-driver/lib-extend.nix:
```diff
@@ -0,0 +1,27 @@
+      })
+    ];
+  };
+  inherit (pkgs.testers) nixosTest;
+  sampleEvaluation = nixosTest {
```
```suggestion
  inherit (pkgs.testers) runNixOSTest;
  sampleEvaluation = runNixOSTest {
```

No awkward `callPackage` dependency injection, and this one supports `imports` and any other module syntax. It also pretty much guarantees that the `pkgs` in the nodes is going to be equal to the `pkgs` passed and catches mistaken `nixpkgs.*` definitions.

I guess I'll improve the docs...

baloo replied:
well found the same bug in `runNixOSTest` then :D

baloo replied:
I'm not comfortable using only `runNixOSTest` as the bugs currently happens when either are used and I need both to be fixed ideally.
I adjusted the tests accordingly, let me know what you think.

chreekat commented:
> @chreekat I'm still confused by your 'code looks different than the description". could you review the first commit first, then the second one.

1. In the first commit, you say "adds support for lib.extend", although it looks like you're actually supporting doing any and everything to `lib`, including replacing it altogether. :) You seem to have a use case that uses `lib.extend`, but that's just a side effect of the change, right?


2. In the first commit's title, you use the present tense to describe the change, which I think is a good practice. But then you go on to write a sentence---also in the present tense---that seems to describe how things used to be before the commit? Or is it providing further details about the effects of your change? I am not sure of your meaning there, so I am not sure what your intended change is.

baloo committed:
runNixOSTest: adds support for lib.extend

When lib overrides were used, before this commit, they would not be made
available in the configuration evaluation of nixosTest's nodes.

Sample code:
``` nix
let
  pkgs = import ./. {
    overlays = [
      (new: old: {
        lib = old.lib.extend (self: super: {
          sorry_dave = builtins.trace "There are no pod bay doors" "sorry dave";
        });
      })
    ];
  };
in
pkgs.testers.runNixOSTest {
  name = "demo lib overlay";

  nodes = {
    machine = { lib, ... }: {
      environment.etc."got-lib-overlay".text = lib.sorry_dave;
    };
  };

  testScript = { nodes }:
    ''
      start_all()
      machine.succeed('grep dave /etc/got-lib-overlay')
    '';
}
```

baloo commented:
@ofborg test nixos-test-driver.lib-extend

baloo commented:
@ofborg test nixos-test-driver.lib-extend

roberth reviewed (commented):
The test wasn't run anymore, nor does it need to be, so I propose to clean that up.
After that, lgtm!

roberth commented:
lgtm, except for the last commit message. Could you squash the last two commits into a single commit?

baloo commented:
@ofborg test nixos-test-driver.lib-extend

roberth commented:
Thanks!
It may not be what I like, but I like that it works.
